### DISCUSSION TITLE
CMSIS-NN on Cortex A72/ Cortex -R5F /C7X-DSP

### DISCUSSION THREAD

**Discussion by prabhatsharmaa** [ARM-software/CMSIS_5#1553] [Q&A]
Hello,

I understand that CMSIS-NN is built for Cortex-M. I would like to understand if it is possible to run CMSIS-NN on **Cortex A72/ Cortex -R5F /C7X-DSP** by cross compilation or by any other method?

I have a controller which doesn't have a Cortex M core. So exploring option to run CMSIS-NN in available cores.

**Comment by JonatanAntoni** [comment]
@felix-johnny et al, can you comment on this one, please?

**Comment by felix-johnny** [comment]
@prabhatsharmaa . Thanks for reaching out. 

CMSIS-NN (usually)has three sections in each function. 
1. pure-C
2.  code with SIMD intrinsics from DSP extension
3. Code with SIMD intrinsics using Helium tech

Technically you could choose to use options 1 and 2 with an Arm Cortex-A CPU..You would have to fix certain custom intrinsics that are defined in CMSIS-Core if using option 2. With the Cortex-A72, using the reference code(C) might be a better option as the compiler can then possibly generate  vector instructions for some of the core operations. For Arm Cortex-R CPU you should be able to compile using option 1. I am not sure about option 2. 

That said, CMSIS-NN is a library, so you'll need a framework like TensorFlow for Microcontroller(TFLM) where this is integrated. I don't think TFLM has support for Cortex-A CPUs in their upstream.
